### ISSUE TITLE
Build out refund API for processing / recording refunds 

### ISSUE THREAD

**Issue by zackkatz** [easydigitaldownloads/easy-digital-downloads#2721]
It would be helpful to see a visual of refunds as well as Earnings and Sales.

**Comment by pippinsplugins** [comment]
This has to be punted for now since we don't yet store the refunded amount. We're already running an upgrade routine for 2.3 and I'd rather not run multiple/

**Comment by pippinsplugins** [comment]
Related to #2625

**Comment by pippinsplugins** [comment]
We're getting much closer on this with #4403 and the recent changes to `EDD_Payment` for refund processing.

Before we can go 100% with this, we need to work out a way to refund line items and/or full amounts.

**Comment by cklosowski** [comment]
I think in order to do this properly, we need our custom tables for Payments. Without the custom tables it's extremely difficult to make these relational information changes with full confidence. Because of that I'm going to have to punt. I want this to be done, but with our current data structure, it will be neither easy or consistant because of our current meta.

Thoughts on this for 2.8/9 @pippinsplugins?

**Comment by pippinsplugins** [comment]
I agree with that.

I hope we can do it in 2.8 or 2.9 but it will depend entirely on the custom tables project.

**Comment by JJJ** [comment]
I'd like for us to tackle this in 3.1, after all of the dust from moving to custom tables has settled.

We are going to put 3.0 beta 1 out today. It's _possible_ that we can sneak this in during beta, but I don't want to set that expectation by having this issue in that milestone until we know how beta goes.

**Comment by pippinsplugins** [comment]
I am okay with this not being in 3.0 so long as it is a top priority for 3.1. It's been one of our longest-standing weak points.

**Comment by JJJ** [comment]
Bringing this back to 3.0.

We'll need this in order to provide a UI for order adjustments anyways. I've already started on this, so this will make it into 3.0. <3

**Comment by JJJ** [comment]
Second pass at a UI that provides room to grow for partials, in `release/3.0` now:

<img width="1232" alt="screen shot 2018-06-08 at 17 01 14" src="https://user-images.githubusercontent.com/88951/41182958-9f6bd640-6b3d-11e8-97ba-8a46b7857459.png">

**Comment by JJJ** [comment]
Things to do here:

* Quick Add/Edit JS
* Non JS Add/Edit pages
* Settle on bulk actions & row actions
* Test with different Adjustment types & depths
* Test with add-ons, more migrated data, etc...

**Comment by cklosowski** [comment]
Just had a conversation with @JJJ about where we should head with this and we've settled on a modal that presents itself.

Since our minimum version is 4.7, we can't rely on the 5.0 dialog JS, so we'll use jQuery Dialog to do this. We had discussed using the Free Downloads modal however, it presents it's own challenges, so using a known library for it is probably best.

We've made some improvements to the styles for it, so we can fine tune that so we don't end up with the stock designs. As I work up a UI for this, i'll post progress images.

**Comment by cklosowski** [comment]
Initial work is done here. Need to do some UI tweaks to make it a little more friendly but here is a preview.

![simplescreenrecorder20190205](https://user-images.githubusercontent.com/1390124/52309797-79e18c00-295e-11e9-9e55-5ca21765e46c.gif)

Things to do still:
- [ ] More fault tolerance code to catch exceptions
- [ ] Include/Verify nonce usage
- [ ] Final UI tweaks
- [ ] Work up gateway support for partial refunds (if they support, and when they do, how we send the partial refund to the gateway)

Stretch Goals:
- [ ] Include adjustments and discounts application to existing orders

**Comment by cklosowski** [comment]
Also fixed  up a bug in WP core when loading a list table via AJAX, and changed the refund item list table to have it's own single and plural definitions.

**Comment by DrewAPicture** [comment]
@cklosowski Is there a PR or branch we could create a PR from with these changes?

**Comment by cklosowski** [comment]
@DrewAPicture I'll have to get one updated with all the recent changes.

**Comment by cklosowski** [comment]
@DrewAPicture PR Created.

**Comment by cklosowski** [comment]
Closing this out. Any changes needed to the refund process/API for gateways to support this new method/concept, should be put into new issues and developed in an isolated scope. This includes support for partial refunds.
